route-rs-packets/src/ethernet.rs

```diff
@@ -0,0 +1,169 @@
+use crate::*;
+use std::borrow::Cow;
+use std::convert::{TryFrom, TryInto};
+
+pub struct EthernetFrame<'packet> {
+    pub data: PacketData<'packet>,
+    pub payload_offset: usize,
+}
+
+impl<'frame> EthernetFrame<'frame> {
+    pub fn new(frame: PacketData) -> Result<EthernetFrame, &'static str> {
+        // Ethernet II frames must be at least the header, which is 14bytes
+        // 0                    6                    12                      14
+        // |---6 byte Dest_MAC--|---6 byte Src_MAC---|--2 Byte Payload Len---|
```

Comment:
In Ethernet II, the 12th and 13th bytes are actually the ethertype field, not a payload length. I believe there is no length field in ethernet II. We will probably need something like the IpProtocol types for EtherTypes as well

Reply:
Interesting.  https://en.wikipedia.org/wiki/Ethernet_frame Looks like it is EtherType in the case it is an Ethernet II frame. As far as I'm aware, we really only expect Ethernet II types on the internet?  https://en.wikipedia.org/wiki/EtherType Looks like it is fairly complex.

Reply:
EtherTypes are how we define what the next packet type is, though -- since LLDP and ARP don't use IP headers. I'm not sure we can avoid it. According to https://en.wikipedia.org/wiki/Ethernet_frame#Types, it looks like Ethernet II is the most common, with raw IEEE 802.3 being an ancient Novell thing.

Reply:
I would honestly just drop support for the 802.3 stuff entirely and just support Ethernet II with the EtherType field. I don't think it's worth it to be generic across so many different standards when we'll never encounter them.

Reply:
We chatted offline about this, my next push will shift us to just supporting EtherType instead of length field. I will still check that it is in a valid EtherType range.